Capture Unhandled errors and add rerun check listener

marcos-arranz opened the pull request:
This captures unhandled YAML parsing errors. On error it returns an `Either` object, which later on is inspected to populate the validation errors

Testing on private repo:

<img width="789" alt="Screenshot 2022-12-05 at 17 13 17" src="https://user-images.githubusercontent.com/103436398/205699877-df1b60ca-531c-4111-a06f-4d04322f3320.png">

This PR also allows to re run a failed checked.

marcos-arranz commented on .yarnrc.yml:
```diff
@@ -0,0 +1 @@
+nodeLinker: node-modules
```
This is added by tool volta which I installed for another project. Do you think it can be added to the project @jsfr @andersfischernielsen ? I have noticed the yarn.lock project was also modified, that's the reason I included it

andersfischernielsen replied:
Are you using Yarn 2? 👀 That'd be the reason this is here afaict 🤔 

But it [looks fine to add](https://yarnpkg.com/configuration/yarnrc#nodeLinker)

dpotyralski reviewed (approved):
LGTM!

andersfischernielsen commented on package.json:
```diff
@@ -26,6 +26,7 @@
     "dotenv": "16.0.3",
     "genson-js": "0.0.8",
     "jszip": "3.10.1",
+    "monet": "^0.9.3",
```
What's the reason for going with `monet` instead of the battle-tested `fp-ts` here?

marcos-arranz replied:
No particular reason as far as I know, @dpotyralski and I talked about using more `Either` than exceptions and this was the library @dpotyralski proposed. Do you recomend move to `fp-ts`?

dpotyralski replied:
> No particular reason as far as I know, @dpotyralski and I talked about using more `Either` than exceptions and this was the library @dpotyralski proposed. Do you recomend move to `fp-ts`?

TBH I don't recall proposing a concrete library, it was more of a concept proposal :)

marcos-arranz replied:
may be, I don’t remember. In any case, I can move it to `fp-ts` if this is a better option

andersfischernielsen replied:
`fp-ts` provides a ton of nice helpers for functional programming, so it'd be preferable to use. It's the most complete FP library out there 😄

andersfischernielsen replied:
`chain` and `pipe` especially for handling the internal `Either` results

andersfischernielsen commented on src/app.ts:
```diff
@@ -53,45 +55,60 @@ const extractPullRequestInformation = (payload: PullRequestEvent) => {
+
+      return validatedTemplates.errors.concat(validatedFiles.errors)
+    } catch (error) {
+      if (error instanceof VersionNotFoundError) {
+        log.debug('Version for %s/%s:%s has not been found', error.owner, error.name, error.version)
+        const validatedTemplate = validateTemplateConfiguration(configurationChanges, getDefaultSchema())
+        return validatedTemplate.errors.concat({
+          message: `Version could not be found at ${error.owner}/${error.repo}:${error.version}`,
+          line: undefined,
+        })
+      } else {
+        throw error
+      }
```
Shouldn't we convert this to a function returning an `Either` instead of using both exception handling/passing _and_ the functional `Either` design pattern?

andersfischernielsen replied:
Traditionally it's either the functional programming pattern (using pure functions returning `Either` results) _or_ exceptions. Mixing both of them in one code base seems risky and could be confusing for readers IMO

dpotyralski replied:
> We had this discussion @dpotyralski and I [here](https://github.com/pleo-oss/file-distributor/pull/73#discussion_r1022929714). In an ideal scenario probably yes, but using `VersionNotFound` with `Either` implies chagning several functions. I am not particularly very annoyed by having both patterns in the same codebase, but let me know your thoughts.

TBH I don't recall proposing a concrete library, it was more of a concept proposal :)

marcos-arranz commented on test/configuration.test.ts:
```diff
@@ -35,9 +37,41 @@ values:
 
-    expect(result.repositoryConfiguration).toEqual(expected)
+    expect(E.isLeft(result)).toBeTruthy()
+    if (E.isLeft(result)) {
```
Needed as after this check is possible to use `.left`. Otherwise, TS complains about it.

andersfischernielsen commented on src/app.ts:
```diff
@@ -50,104 +49,42 @@ const extractPullRequestInformation = (payload: PullRequestEvent) => {
-    const approvedReviewId = await approvePullRequestChanges(repository, prNumber)
-    log.debug(`Approved PR #%d in %s.`, prNumber, approvedReviewId)
-  }
+  await processCheckRun(enrichedWithRepoLog, octokit, configFileName, prNumber, repository, sha, checkId)
```
I think we should use the same pattern here for the functions that have been extracted as in other modules. That way we don't pass 7 parameters in a single function and can instead encapsulate the logger and Octokit like so:

```suggestion
//Outside of this function
import { validation } from './core-validation'
  //...
  const { processCheckRun } = validation(enrichedLogger, octokit)
  await processCheckRun(configFileName, prNumber, repository, sha, checkId)
```

I still think 5 parameters is too much (does the function do more than one thing or can we combine the data?), but that can be fixed later

andersfischernielsen commented on src/app.ts:
```diff
@@ -50,104 +49,42 @@ const extractPullRequestInformation = (payload: PullRequestEvent) => {
+const processPullRequest = async (payload: PullRequestEvent, context: Context<'pull_request'>) => {
+  const { number: prNumber, sha, repository } = extractPullRequestInformation(payload)
+  const { log, octokit } = context
+  const enrichedWithRepoLog = log.child({ owner: repository.owner, repository: repository.repo })
 
-  log.info(`Validated configuration changes in #%d with conclusion: %s.`, prNumber, checkConclusion)
+  await processCheckRun(enrichedWithRepoLog, octokit, configFileName, prNumber, repository, sha)
```
Same feedback as https://github.com/pleo-oss/file-distributor/pull/80/files#r1043298474 for this use of the new function.

andersfischernielsen commented on src/app.ts:
```diff
@@ -205,6 +143,10 @@ export = async (app: Probot) => {
     await processPushEvent(context.payload as PushEvent, context)
   })
 
+  app.on('check_run.rerequested', async (context: Context<'check_run'>) => {
+    await processCheckRerun(context.payload as CheckRunRerequestedEvent, context)
+  })
```
Is this part of the bugfix or new functionality? I'm slightly confused why this is necessary

marcos-arranz replied:
Part of the new functionality. It is necessary because if the check fails because a transient error, like a network error, the user probably wants to re run the check. If we don't implement this then the button on the checks won't do anything.

andersfischernielsen commented on src/checks.ts:
```diff
@@ -27,9 +27,35 @@ export const checks = (log: Logger, octokit: Pick<OctokitInstance, 'checks'>) =>
     const { repo, owner, checkRunId, sha, conclusion } = input
 
-    if (input.errors.length > 0) {
+    if (
+      input.checkResolution === CheckResultion.ERROR ||
+      (input.checkResolution === CheckResultion.FAILURE && input.errors.length === 0)
+    ) {
```
Couldn't we use the existing type that GitHub provides here? We can always make is easier to read and pass around by defining a 

```typescript
export type CheckConclusion =
  | 'action_required'
  | 'cancelled'
  | 'failure'
  | 'neutral'
  | 'success'
  | 'skipped'
  | 'timed_out'
  | null
  ```
  type 😄

marcos-arranz replied:
Unfortunately Github doesn't seem to offer a type for that field:

![Screenshot 2022-12-09 at 10 33 41](https://user-images.githubusercontent.com/103436398/206682789-fda7dbfa-c8d5-4507-8c79-d5ab40655980.png)

I can define our own.

andersfischernielsen commented on src/checks.ts:
```diff
@@ -27,9 +27,35 @@ export const checks = (log: Logger, octokit: Pick<OctokitInstance, 'checks'>) =>
+      input.checkResolution === CheckResultion.ERROR ||
+      (input.checkResolution === CheckResultion.FAILURE && input.errors.length === 0)
+    ) {
+      const {
+        data: { conclusion: result },
+      } = await octokit.checks.update({
+        owner,
+        repo,
+        name: 'Configuration validation',
+        check_run_id: checkRunId,
+        status: 'completed',
+        head_sha: sha,
+        conclusion,
+        output: {
+          title: 'Schema validation',
+          summary:
+            'There was an unexpected error running the check. Please try again and if the error persists contact the stewards.',
+        },
```
I'd extract the `if` and `else` out here to separate contained functions. It's really hard for me to understand everything that's happening in these conditionals

andersfischernielsen commented on src/configuration.ts:
```diff
@@ -45,35 +50,42 @@ export const configuration = (log: Logger, octokit: Pick<OctokitInstance, 'repos
-      repositoryConfiguration: combinedConfiguration,
-      cstYamlRepresentation: rep,
-    } as TemplateConfig
+      return E.right({
+        repositoryConfiguration: combinedConfiguration,
+        cstYamlRepresentation: rep,
+      } as TemplateConfig)
+    } catch (error) {
+      if (error instanceof YAMLParseError) {
+        return E.left(error)
+      }
+      throw error
+    }
```
Why are we still catching and rethrowing exceptions? The idea of `Either` is to map every result or failure to a consistent type-safe structure so we avoid having any `try/catch` across the code 🤔

andersfischernielsen replied:
Instead of `parse`, we can use `parseDocument` which doesn't throw on any YAML errors. Then we can just return `document.errors` if there are any `YAMLError`s 😄

marcos-arranz replied:
> Instead of parse, we can use parseDocument which doesn't throw on any YAML errors. Then we can just return document.errors if there are any YAMLErrors 😄

`parseDocument` seems to return a different type that cannot be safely casted to `RepositoryConfiguration`

andersfischernielsen commented on src/core-validation.ts:
```diff
@@ -0,0 +1,159 @@
+import { ProbotOctokit } from 'probot'
+import { git } from './git'
+
+const validateChanges = async (
+  log: Logger,
+  octokit: Pick<OctokitInstance, 'repos'>,
+  configurationChangesOrError: E.Either<YAMLParseError, TemplateConfig>,
+): Promise<ValidationError[]> => {
```
I'd prefer (see https://github.com/pleo-oss/file-distributor/pull/80/files#r1043298474) if we could keep the modules in the app consistent 😄 That way it's easier to understand every module when the structure is the same. 

Functions that take either a logger or an Octokit can be part of a function that takes the logger and Octokit and injects them into the functions that use them. These functions are then returned as a function providing:

```typescript
return {
  validateChanges,
  processCheckRun
}
```

That way we limit the parameter passing and a ton of parameters for every function 😄

marcos-arranz replied:
Addressed

andersfischernielsen commented on src/core-validation.ts:
```diff
@@ -0,0 +1,159 @@
+    )
+
+    log.info(`Validated configuration changes in #%d with conclusion: %s.`, prNumber, checkConclusion)
+  } catch (error) {
+    await resolveCheckRun(
+      {
+        ...checkInput,
+        conclusion: 'failure',
+        checkRunId: checkId,
+        errors: [],
+        checkResolution: CheckResultion.ERROR,
+      },
+      configFileName,
+    )
+    throw error
```
Why are we throwing if we're returning an `Either`? 🤔

andersfischernielsen commented on src/types.ts:
```diff
@@ -63,10 +63,17 @@ export interface CreateCheckInput {
   sha: string
 }
 
+export enum CheckResultion {
+  ERROR,
+  FAILURE,
+  SUCCESS,
+}
```
These could be bound to the `'action_required' | 'cancelled' | 'failure' | 'neutral' | 'success' | 'skipped' | 'timed_out' | null` type provided by GitHub (if we _really_ want an enum here, I doubt it) or be converted to a type exactly like GitHub's to avoid having to map from their type to ours 😄

marcos-arranz replied:
I am not able to find that type. Can you point me to the definition?

andersfischernielsen commented on src/types.ts:
```diff
@@ -63,10 +63,17 @@ export interface CreateCheckInput {
   conclusion: string
   checkRunId: number
   errors: ValidationError[]
+  checkResolution: CheckResultion
```
Isn't this the same as `conclusion`? 🤔

marcos-arranz replied:
It is not exactly the same, `CheckResolution` can be `ERROR` or `FAILURE`, the former means that there was an error due to the application, and the later that the input given by the user was wrong. This result in one saying something like "There was an unhandled error" and other saying something like "There is an error in your yaml, please fix it". On both of them `conclusion` is `failure`

I can try to use just conclusion and report it as `action_required` for failure. WDYT?

andersfischernielsen commented on src/types.ts:
```diff
@@ -63,10 +63,17 @@ export interface CreateCheckInput {
   sha: string
 }
 
+export enum CheckResultion {
```
```suggestion
export enum CheckResolution {
```

andersfischernielsen commented on test/app.test.ts:
```diff
@@ -106,11 +104,11 @@ describe('Probot Tests', () => {
     }
 
     await probot.receive(pushEvent as unknown as EmitterWebhookEvent)
+    expect(baseNock.isDone()).toBeTruthy()
```
What does this do? 🤔  Shouldn't Nock already explode by itself for any unresolved mocks?

marcos-arranz replied:
No, it does not explode without this.

andersfischernielsen commented on package.json:
```diff
@@ -55,5 +56,9 @@
     "ts-jest": "29.0.3",
     "typescript": "4.8.4"
+  },
+  "volta": {
+    "node": "18.12.1",
+    "yarn": "1.22.19"
```
Why do we need the Volta configuration again? None of the project relies on Volta for building and distribution 🤔

marcos-arranz replied:
I installed as per instructions in Grim Reaper https://github.com/pleo-io/grim-reaper/blob/main/package.json#L65. I haven't found any way to tell volta to pin yarn without modifying `package.json`

andersfischernielsen commented:
Maybe we should split the bugfix into a `patch` PR and add the additional functionality in a `minor` PR. That'd also be making it easier to handle the module, parameter and structure comments, if there are any 😄

marcos-arranz commented:
I think it makes sense to split it @andersfischernielsen , this PR is quite big and has two different purposes. I am going to address your comments here and send a different PR for the YAML error, so comments are preserved

marcos-arranz commented on src/app.ts:
```diff
@@ -170,9 +107,10 @@ const processPushEvent = async (payload: PushEvent, context: Context<'push'>) =>
   if (!filesChanged.includes(configFileName)) return
 
-  const parsed = await determineConfigurationChanges(configFileName, repository, payload.after)
+  const either = await determineConfigurationChanges(configFileName, repository, payload.after)
```
`parsed` is used below alreay. I can change `either` to `errorOrParsed`. WDYT?

marcos-arranz replied:
Or use pipe as in 

```
return pipe(
    either,
    E.match(
      () => undefined,
      async (parsed: TemplateConfig) => {
        const { configuration: defaultValues } = await getTemplateInformation(parsed.repositoryConfiguration.version)

        const combined = combineConfigurations(defaultValues, parsed.repositoryConfiguration)
        if (!combined) return

        const { version, templates: processed } = await renderTemplates(combined)
        const pullRequestNumber = await commitFilesToPR(repository, version, processed)

        if (!pullRequestNumber) {
          log.info('Commit leads to no changes.')
          log.info('Skipped PR creation.')
        }

        log.info('Committed templates to %s/%s in #%d', repository.owner, repository.repo, pullRequestNumber)
        log.info('See: https://github.com/%s/%s/pull/%d', repository.owner, repository.repo, pullRequestNumber)
      },
    ),
  )
```

but I feel it's a bit overkill

marcos-arranz replied:
Also, I am not sure how it works with the `async` modifier

marcos-arranz commented:
@dpotyralski @andersfischernielsen I have included the changes from the other PR and addressed your comments. Let me know if you have any other concern, otherwise I can merge this.

andersfischernielsen commented on src/core-validation.ts:
```diff
@@ -0,0 +1,147 @@
+
+      const onlyChangesConfiguration = filesChanged.length === 1 && filesChanged[0] === configFileName
+
+      if (errors.length > 0) {
+        const changeRequestId = await requestPullRequestChanges(repository, prNumber, checkId)
+        log.debug(`Requested changes for PR #%d in %s.`, prNumber, changeRequestId)
+      } else if (onlyChangesConfiguration) {
+        const approvedReviewId = await approvePullRequestChanges(repository, prNumber)
+        log.debug(`Approved PR #%d in %s.`, prNumber, approvedReviewId)
+      }
```
This was patched out on `main` 😬

andersfischernielsen replied:
Specifically #85 and #87
